### ISSUE TITLE
unknown keywords for ssl after compiling with support for openssl

### ISSUE THREAD

**Issue by LightAutumnMelancholy** [haproxy/haproxy#752]
removed

**Comment by wtarreau** [comment]
Just to clarify, you mean that you've declared a server with the "sni" keyword for example and without the "ssl" keyword, and that you got an error about the "sni" keyword ? I'm asking to clarify because that situation seems odd to me, the keywords should be ignored since "ssl" is not present.

**Comment by DarklyAutumnClad** [comment]
Ah, 

So the message "unknown keywords for ssl in backend" appears if you have defined a port and 'ssl' in a backend. I am hoping that you'll add an error message that says "ssl and port cannot be defined at the same time" or something to that effect. I originally thought that SNI was dependent on the same, but may just have been depended on ssl keyword to be present. 

At any rate, top results all lead me to believe/try recompiling, even if support was compiled (I explicitly compiled with SSL, USE_OPENSSL=1 was originally done.)

As you go through troubleshooting and research you find that error messages that contain "unknown keywords ssl in backend" often lead folks to recompile with with SSL_INC, and SSL_LIB, as they end up suspecting that the SSL support was not fully compiled in.

The problem as I see it is a problem of an error message that could be improved: 

"unknown keyword ssl in backend" should refer to a situation where support was not compiled for it. Instead of getting that error if the ssl keyword and a custom :int for a port have been defined. 

I wasted several hours having missed the one sentence in the backend section of the manual on ssl. Where port number and ssl cannot be defined together.

**Comment by lukastribus** [comment]
I'm not sure I understand what you are talking about in the first place, could you please just post an *actual example* of a backend configuration that leads to this specific error message?

**Comment by DarklyAutumnClad** [comment]
@lukastribus  I'm doing my best to explain, so let me reduce this, so that it can be more easily understood. 

*  "_unknown keyword ssl in backend_" is an error message for when the server section has been misconfigured with both 'ssl' and ':port' 

* You cannot configure both.  "_unknown keyword ssl in backend_" is associated with a failure to compile support for OpenSSL.

I think it would be optimal to check if both are configured, and raise an error that simply states: "Do not configure keyword 'ssl' and port number, it's unsupported" , so that administrators can quickly and easily identify the problem.

I did not think showing
 
```
backend 
   server domain.com:443 ssl
```

Will/would help understand why the error message is suboptimal. The error message is sub-optimal, because it leads an admin to believe that there is a problem with compilation, and not just that you cannot do both of these: ":1234 ssl".  It would be optimal to instruct admins not to do this in the future.

Forgive me if that was incorrect, but this is a suggestion. Thank you for your time. My apologies for any confusion.

**Comment by wtarreau** [comment]
So I'm sorry but even after copying (and fixing) your example config, I cannot reproduce your error:
```
backend f
   server s domain.com:443 ssl

[WARNING] 199/212721 (25224) : config : missing timeouts for backend 'f'.
   | While not properly invalid, you will certainly encounter various problems
   | with such a configuration. To fix this, please ensure that all following
   | timeouts are set to a non-zero value: 'client', 'connect', 'server'.
[NOTICE] 199/212721 (25224) : haproxy version is 2.3-dev0-273aea-59
[NOTICE] 199/212721 (25224) : path to executable is ./haproxy
[ALERT] 199/212721 (25224) : Proxy 'f', server 's' [err.cfg:2] verify is enabled by default but no CA file specified. If you're running on a LAN where you're certain to trust the server's certificate, please set an explicit 'verify none' statement on the 'server' line, or use 'ssl-server-verify none' in the global section to disable server-side verifications by default.
[ALERT] 199/212721 (25224) : Fatal errors found in configuration.
```

Could you please once for all post a config that reproduces the issue you're seeing ? Your explanation is far from being clear and the examples you give unsurprisingly don't work.

**Comment by lukastribus** [comment]
The syntax is:

`server <servername> <hostname/ip>[:<port>] [ssl]`

Specifying the port together with the `ssl` keyword is absolutely supported, in fact it's extremely uncommon to NOT specify the port.

It works in haproxy 1.5 and in the current development tree.


When you forget to specify the servername, like in your example above:

haproxy 1.5 returns:

    [ALERT] 199/220856 (6570) : parsing [../cert/sslconfig-example.cfg:44] : 'server domain.com:443' : invalid address: 'ssl' in 'ssl'


haproxy (current) returns:

    [ALERT] 199/220329 (6566) : parsing [../cert/sslconfig-example.cfg:44] : 'server domain.com:443' : could not resolve address 'ssl'.


Is that maybe something that happened here? If not, please give us the real example with the real error message.

**Comment by DarklyAutumnClad** [comment]
I'll see what my employer will let me release, I'll need to get permission and sanitize a config to show. (if you are okay with a sanitized config) You'll hear back from me Monday most likely. I apologize, I thought I was really just going to make a quick suggestion here, and not fall down the rabbit hole with it. I'll get some configs to show. 

I understand that this is not the expected outcome for this based on what you are showing me, so likely config, must be a problem when used in conjunction. I suspect something in globals, probably a check value.

I tested against multiple versions with the same configuration and got this error every time, among others that were the result of changes across versions. 

The result of doing this (with and without resolvers)

```
backend
   server foo bar.com:443 ssl
```
Resulted in 

Unknown keyword 'ssl' in backend

Only when I removed bar.com's :443 did this boot with success.

haproxy failed startup on 1.8.25, 19.9 and a 2.0.x that I tried.

I found this was in the instructions for 1.8+, and I checked to see if it was present against other versions, which it seemed to be.

"This option enables SSL ciphering on outgoing connections to the server. It
is critical to verify server certificates using "verify" when using SSL to
connect to servers, otherwise the communication is prone to trivial man in
the-middle attacks rendering SSL useless. When this option is used, health
checks are automatically sent in SSL too unless there is a "port" or an
"addr" directive indicating the check should be sent to a different location.
See the "no-ssl" to disable "ssl" option and "check-ssl" option to force
SSL health checks."

http://cbonte.github.io/haproxy-dconv/1.8/configuration.html#5.2-ssl
http://cbonte.github.io/haproxy-dconv/1.9/configuration.html#5.2-ssl
http://cbonte.github.io/haproxy-dconv/2.0/configuration.html#5.2-ssl
http://cbonte.github.io/haproxy-dconv/2.1/configuration.html#5.2-ssl
http://cbonte.github.io/haproxy-dconv/2.2/configuration.html#5.2-ssl

This is why I tried removing the :port, and then did boot with success.

**Comment by wtarreau** [comment]
A sanitized config is indeed fine, but even if you're able to simplify a piece of configuration to the minimal required parts, this will save you from having to disclose anything internal, and will make it easier for us to spot the root cause of the problem. Have a nice week-end!

**Comment by DarklyAutumnClad** [comment]
`##############################
# GlobalOpts
##############################
global
    log         1007.0.0.1 local2 info
    chroot      /var/lib/haproxy
    maxconn     4000
    user        haproxy
    group       haproxy
    daemon
    master-worker
    stats socket /var/lib/haproxy/stats mode 777 expose-fd listeners level admin

    ssl-default-bind-ciphers redacted
    ssl-default-server-ciphers redacted


##############################
# Common defaults
##############################
defaults
    mode                    http
    log                     global
    option                  httplog
    option                  http-ignore-probes
    option                  forwardfor
    option                  splice-auto
    option                  redispatch
    retries                 3
    timeout http-request    60s
    timeout queue           1000s
    timeout connect         20s
    timeout foo          1000s
    timeout server          60s
    timeout http-keep-alive 20s
    timeout check           20s
    maxconn                 2000

##############################
# dns 
##############################
resolvers cloudflare
    nameserver dns1 1.1.1.1:53
    nameserver dns2 1.0.0.1:53
    resolve_retries 3
    timeout retry 1s
    hold valid 20s

##############################
# main 
##############################
frontend front
    bind *:80
    redirect scheme https if { hdr(host) -i bar-test-borked.baz.com } !{ ssl_fc }
    bind *:443 ssl crt /etc/ssl/certs/gz/baz.com.pem
    http-response set-header Strict-Transport-Security max-age=15768000
    capture request header Host len 32
    # ACL:
    acl is_myget_calling hdr('xdps)' -i True path_beg /api/v2/RetrieveItems
    acl is_myget_bound path_beg /api/v2/RetrieveItems unless hdr('xdps')
    acl is_myget_calling hdr('xdps)' -i True path_beg /api/V2/RetrieveItems
    acl is_myget_bound path_beg /api/V2/RetrieveItems unless hdr('xdps')
    acl is_crfa path_beg /api/V2/Getcrfa
    acl safe_ips src 0.0.0.0

    acl is_healthy path_end -i /api/healthy
    acl is_foo hdr(host) -i fooapi-bar.baz.com
    acl is_occurence path_beg /api/V2/checkoccurence
    acl is_frontdoor hdr(host) -i bar-test-borked.baz.com
    acl is_proxy hdr(host) -i bar-test-haproxy.baz.com

    # USE:
    use_backend occurence if is_occurence
    use_backend commander if is_crfa
    use_backend myget if is_myget_bound !is_myget_calling
    use_backend borke if is_myget_calling is_myget_bound
    use_backend ishealthy if is_healthy
    http-request silent-drop if is_foo !{ path_beg -i /api /recieve_itemsApi.srv }
    use_backend borke if is_foo
    http-request deny if is_frontdoor !safe_ips

    use_backend frontdoor if is_frontdoor
    use_backend stats if is_proxy

##############################
# Secondary 
##############################
frontend frontdoor
    bind 20.100.100.252:80
    redirect scheme https if { hdr(host) -i bar-test-borked.baz.com } !{ ssl_fc }
    bind 20.100.100.252:443 ssl crt /etc/ssl/certs/gz/baz.com.pem
    http-response set-header Strict-Transport-Security max-age=15768000

    # ACL:
    acl safe_ips src 0.0.0.0
    acl is_frontdoor hdr(host) -i bar-test-borked.baz.com
    acl is_proxy hdr(host) -i bar-test-haproxy.baz.com
    acl is_healthy path_end -i /api/healthy
    http-request deny if is_frontdoor !safe_ips

    # USE:
    use_backend ishealthy if is_healthy
    use_backend frontdoor if is_frontdoor
    use_backend stats if is_proxy

##############################
# Borke backend
##############################
backend borke
    fullconn 800
    log global
    default-server inter 3s fall 3 rise 2
    balance roundrobin
    option httpchk GET /api/healthy
    acl is_dumpstercat req.hdr(Content-Type) application/enc_url
    http-request set-header Content-Type application/json unless is_dumpstercat
    http-request silent-drop if { path_beg -i /discussions  }

    reqrep ^([^\ :]*)\ /recieve_itemsApi.srv/recieve_items/(.+)                           \1\ /api/foo/v1/recieve_items/\2
    reqrep ^([^\ :]*)\ /api/[vV](?:2|165)/checkoccurence(.*)                      \1\ /api/occurence/v1\2
    reqrep ^([^\ :]*)\ /api/[vV](?:2|165)/recieve_itemsApi.srv/recieve_items/(.+)         \1\ /api/foo/v1/recieve_items/\2
    reqrep ^([^\ :]*)\ /api/[vV](?:2|165)/recieve_itemsApi.srv/recieve_itemsVersion/(.+)  \1\ /api/foo/v1/recieve_items/\2
    reqrep ^([^\ :]*)\ /api/[vV](?:2|165)/checkValidborke(.*)                    \1\ /api/foo/v1/checkValid\2
    reqrep ^([^\ :]*)\ /api/[vV](?:2|165)/RetrieveItems(.*)                           \1\ /api/foo/v1/operations\2
    reqrep ^([^\ :]*)\ /api/[vV](?:2|165)/SignIn(.*)                          \1\ /api/foo/v1/SignIn\2
    reqrep ^([^\ :]*)\ /api/[vV](?:2|165)/AmendDeal(.*)                 \1\ /api/foo/v1/occurences\2
    reqrep ^([^\ :]*)\ /api/[vV](?:2|165)/retDevelopmentManifest(.*)               \1\ /api/foo/v1/folders/config\2
    reqrep ^([^\ :]*)\ /api/[vV](?:2|165)/retDevelopmentOperation(.*)                  \1\ /api/foo/v1/folders/operations\2
    reqrep ^([^\ :]*)\ /api/V1/retDevelopmentManifest                              \1\ /api/foo/v1/folders/config\2
    reqrep ^([^\ :]*)\ /api/V1/retDevelopmentOperation                                 \1\ /api/foo/v1/folders/operations\2
    reqrep ^([^\ :]*)\ /api/[vV](?:2|165)/api\.srv                              \1\ /api/healthy
    reqrep ^([^\ :]*)\ /api/V2/retOperationManifest(.*)                             \1\ /api/foo/v1/app_list_config\2
    reqrep ^([^\ :]*)\ /api/V2/retLocationManifest(.*)                              \1\ /api/foo/v1/locale_config\2

    server 20.100.100.208 20.100.100.208:4000 maxconn 500 check inter 5000
    server 20.100.4.95 20.100.4.95:4000 maxconn 500 check inter 5000

##############################
# frontdoor backend
##############################
backend frontdoor
    fullconn 200
    log global
    option httpchk GET /discussions /new

    server 20.100.4.95 20.100.4.95:4000 maxconn 20 check inter 5000

##############################
# Myget
##############################
backend myget
    log global
    reqrep ^([^\ :]*)\ /api/[vV](?:2|165)/RetrieveItems(.*)                           \1\ /myget/\2
    server myget someplace.execute-api-aws-region.amazonaws.com resolvers cloudflare ssl sni str(someplace.execute-api-aws-region.amazonaws.com) verify none
    http-request set-header host someplace.execute-api-aws-region.amazonaws.com
	
##############################
# Backend healthy
##############################
backend ishealthy
    fullconn 800
    balance roundrobin
    option httpchk GET /api/healthy

    server 20.100.100.208 20.100.100.208:4000 maxconn 20 check inter 5000
    server 20.100.4.95 20.100.4.95:4000 maxconn 20 check inter 5000

##############################
# Occurences
##############################
backend occurence
    fullconn 500
    log global
    option httpchk GET /version
    reqrep ^([^\ :]*)\ /api/[vV](?:2|165)/checkoccurence(.*)                      \1\ /api/occurence/v1\2

    server occurence occurence-fu-faf-1.load.baz.com:80 check resolvers cloudflare

##############################
# Commander
##############################
backend commander
    fullconn 200
    log global
    option httpchk GET /version
    reqrep ^([^\ :]*)\ /api/V2/Getcrfa(.*)                                      \1\ /api/SignIn\2

    server commander commander-fu-faf-1.load.baz.com:80 check resolvers cloudflare

##############################
# Stats
##############################
listen stats
    option  dontlognull
    option  dontlog-normal
    bind    0.0.0.0:8080
    stats   enable
    stats   uri /
    stats   hide-version
    stats   auth adminuser:somepassword
`

**Comment by TimWolla** [comment]
After I took the time to fix up the fatal errors in your given configuration I am still unable to reproduce the issue. Next time provide an actual *minimal* configuration that config checks without any errors as well as the full, unredacted, `haproxy -vv` output. Also I strongly recommend you to take care of all the warnings in your configuration.

```
[timwolla@/s/haproxy ((v1.8.25))]grep -A5 '^backend myget' crasher.cfg
backend myget
    log global
    reqrep ^([^\ :]*)\ /api/[vV](?:2|165)/RetrieveItems(.*)                           \1\ /myget/\2
    server myget example.com:443 resolvers cloudflare ssl sni str(someplace.execute-api-aws-region.amazonaws.com) verify none
    http-request set-header host someplace.execute-api-aws-region.amazonaws.com
	
[timwolla@/s/haproxy ((v1.8.25))]./haproxy -c -f ./crasher.cfg
[WARNING] 202/001444 (8375) : parsing [./crasher.cfg:78] : a 'http-request' rule placed after a 'redirect' rule will still be processed before.
[WARNING] 202/001444 (8375) : parsing [./crasher.cfg:80] : a 'http-request' rule placed after a 'redirect' rule will still be processed before.
[WARNING] 202/001444 (8375) : parsing [./crasher.cfg:99] : a 'http-request' rule placed after a 'redirect' rule will still be processed before.
[WARNING] 202/001444 (8375) : parsing [./crasher.cfg:155] : a 'http-request' rule placed after a 'reqxxx' rule will still be processed before.
[WARNING] 202/001444 (8375) : config : missing timeouts for frontend 'front'.
   | While not properly invalid, you will certainly encounter various problems
   | with such a configuration. To fix this, please ensure that all following
   | timeouts are set to a non-zero value: 'client', 'connect', 'server'.
[WARNING] 202/001444 (8375) : config : missing timeouts for frontend 'frontdoor'.
   | While not properly invalid, you will certainly encounter various problems
   | with such a configuration. To fix this, please ensure that all following
   | timeouts are set to a non-zero value: 'client', 'connect', 'server'.
[WARNING] 202/001444 (8375) : config : missing timeouts for proxy 'stats'.
   | While not properly invalid, you will certainly encounter various problems
   | with such a configuration. To fix this, please ensure that all following
   | timeouts are set to a non-zero value: 'client', 'connect', 'server'.
Configuration file is valid
[timwolla@/s/haproxy ((v1.8.25))]./haproxy -v
HA-Proxy version 1.8.25 2020/04/02
Copyright 2000-2020 Willy Tarreau <willy@haproxy.org>
```

**Comment by DarklyAutumnClad** [comment]
@TimWolla  You're being insulting. That configuration is in use for 1.8.25, and that's running on about 60 hosts dedicated to haproxy duties. Those nodes service billions of devices monthly. I'm done with this.

**Comment by wtarreau** [comment]
On Mon, Jul 20, 2020 at 03:42:09PM -0700, DarklyAutumnClad wrote:
> @TimWolla  You're being insulting. That configuration is in use for 1.8.25,
> and that's running on about 60 hosts dedicated to haproxy duties. Those nodes
> service billions of devices monthly. I'm done with this. 

Tim is not insulting, he's trying to help with the issue you're reporting.
YOU are the one saying there is a problem with error reporting, yet YOU
are the only one who manages to see this issue and YOU are the one still
failing to explain how to reproduce it.

When you posted your configuration, Tim noticed that you're having issues
there that can cause you some trouble over time and warns you about them.
If you consider that anyone trying to help you is insulting, then that's
fine, continue to run your prod with your bugs, and don't complain the day
you face a massive outage that you were warned against multiple times by
haproxy itself, then by others trying to help you in good faith.

What is sure is that I wouldn't want my production being handled this way...

Good luck!
Willy

**Comment by LightAutumnMelancholy** [comment]
That config had no fail conditions for boot. Warnings for timeout, are
irrelevant. We handle that at the ELB. We don't need to do it twice. 

The truth is, you'd rather take a shot at something not being
perfect than address the problem, which was described over and over to you.

I didn't appreciate and I will not have someone act passive aggressive with
me online. 

Next time provide an actual minimal configuration that config checks without any errors as well as the full, unredacted, haproxy

Yeah no, I redact that to protect my employer. VV is a waste of time. I told you that OpenSSL was compiled in.


If you think the way you all caps, italicize, and bold is lost on anyone, it's not. 

The fact that you have turned now to personal attacks, legitimizes my
position.

Goodbye and good luck.

**Comment by wtarreau** [comment]
There were no personal attacks, you reported a problem in a way that was hard to translate to a reproducible one, and then when warned that your config triggers warning, you consider that insulting, just as if having a warning in one's config was shameful, and respond in an agressive way to the person trying to help you. You could easily have said "missing timeouts come from the missing parts", "thanks for the warning about the misplaced redirect rules, I didn't notice", etc. But instead you took it personal and decided to close the issue and ignore the problems. It's your call.

**Comment by LightAutumnMelancholy** [comment]
`After I took the time to fix up the fatal errors in your given configuration` 
There were no errors, config is valid. He didn't show at all what he fixed. So I have doubt that he did do that, considering this boots in production on 1.8.25. but he made sure it dripped with contempt.  "After I took the time"  The config is pulled from production and pattern matched and replaced for domain names and uri. 15+ admins work this config, I didn't write it all. This config boots in prod, with 4 of the 6 warnings he provided, which are for ACLs. ACLs which are proven to be good.  

`"Next time provide an actual minimal configuration that config checks without any errors"`

An Actual config was provided. This is public, you don't talk down to people. That's passive aggressive language. Contempt and passive aggressive.


I just don't need that in my life. Best wishes to you all. Thank you for the time, at any rate.

**Comment by wtarreau** [comment]
If that's the way you take it, I'm afraid you'll always have a hard time seeking help from anyone, but maybe that's not what you want anyway. Good luck with your production though, and if any of you 14+ co-admins wants to reopen the issue with exploitable information, they'll be welcome.
PS: you should really take care of your warnings. Just saying.

**Comment by TimWolla** [comment]
> There were no errors, config is valid. He didn't show at all what he fixed. So I have doubt that he did do that, considering this boots in production on 1.8.25.

For posteriority, these are the issues I remember I had to fix from top of my head:

- Replace `=` with `#`, because the configuration you posted used `=` as a comment sign.
- Comment out `timeout foo 1000s`, because there is no `timeout foo` in a stock HAProxy build.
- Fix the IP address in the `log` option within the `global` section, because `1007.0.0.1` is not a valid IP address.
- Specify `ECDHE` within `ssl-default-bind-ciphers` and `ssl-default-server-ciphers` because the values have been replaced by `redacted` (I don't understand why a list of ciphers would be a private thing that needs redacting).
- Replace the hostnames in the `server` section, because they failed to resolve, preventing a successful configuration check.
- Comment out `user` and `group` options, because my workstation does not have the `haproxy` user and group (this one I'm not counting against your configuration).

And I even recompiled HAProxy several times, because I usually test using the current master instead of 1.8 and then I forgot to enable `USE_PCRE`. Not something I am counting against your configuration either, because that clearly is my fault.

It is entirely possible that the configuration works fine in your environment, maybe you are performing some preprocessing step that was missing from the configuration you posted here. In any case the configuration failed to config check successfully *after multiple iterations* of fixing stuff here and there and rerunning a `./haproxy -c -f ./config.cfg`. After you already got several replies from Willy and Lukas I was expecting a minimal working example I can simply copy into a file and immediately reproduce the  issue. Even a non-minimal example would be fine for me if it config checks (or rather fails with the misleading message you reported) on the first attempt.

> Yeah no, I redact that to protect my employer. VV is a waste of time. I told you that OpenSSL was compiled in.

Honest question: Why would `-vv` expose anything about your employer? Are you running custom patches on top of HAProxy?
